apps/server/src/modules/learnroom/service/metadata-loader.service.ts

```diff
@@ -1,15 +1,16 @@
 import { Injectable, NotImplementedException } from '@nestjs/common';
-import { EntityId, LearnroomMetadata, LearnroomTypes } from '@shared/domain';
-import { CourseRepo } from '@shared/repo';
+import { BaseMetadata, Course, EntityId, LearnroomTypes, Lesson } from '@shared/domain';
+import { CourseRepo, LessonRepo } from '@shared/repo';
 
-type RepoType = CourseRepo;
+type RepoType = CourseRepo | LessonRepo;
 
 @Injectable()
 export class MetadataLoader {
```

Comment:
Im not happy with the fact that the learnroom Module is responsible for loading both courses and lessons. If you expand this for tasks, it will get even worse, and severely undermine the borders between modules. I've been looking at how to untangle this, ill try to formulate my suggestion here, feel free to seek me out and we can look at it together if there are problems I overlooked.

First off, this class is currently only used by the share module. So lets ignore other usecases for a bit.
As written in another comment, the connection between ShareTokenParentType and LearnroomType seems to be a mistake (from before your time). I would like to seperate them.

What was correct, is that the learnroom module should be responsible for loading data from courses (learnroom=course atm). however, loading that data from lessons, should be a responsability of the lesson service.
So, we need a service similar to this in the lesson module, which can give us the metadata of a lesson.

That would mean this loader would only need to handle courses specifically, and to be honest, for that its a bit overengineered.

What we need in my opinion is:
- a function in the lessonUc (for now, we can later see if it should be a seperate class) that diffenciates between the different things that can be shared (based on ShareTokenParentType)
- a service in the learnroom module that can be used to get a course. On the surface, we only need the name of that course for our case, so it would be sufficient to return either a new type (like your BaseMetadata) or even the existing learnroomMetadata. Another option would be to simply return the entity. In fact, there already is such a service: the CourseService. So we might as well just use that instead.
- a corresponding service in the lesson module to get a lesson, or its metadata. again there already is such a service: the LessonService.

when we use the CourseService and LessonService, that means the function in the shareuc will know that it will get a course or a lesson. I think thats fine. the function can then simply return the metadata object the shareservice wants, with id, type, and name.

With that, the whole code should be a lot simpler, and then we can look at it again if we want to make further changes.

Reply:
I created this mess in share-token.service.ts :
https://github.com/hpi-schul-cloud/schulcloud-server/blob/e87677775e4067c2416d3b393d948f004daebbd1/apps/server/src/modules/sharing/service/share-token.service.ts#L42-L60
because we need to get both courseName and lesson name. If you have better suggestion please let me know.